allennlp/data/dataset_readers/semantic_parsing/grammar_based_text2sql.py

```diff
@@ -0,0 +1,140 @@
+        self._use_all_sql = use_all_sql
+        self._remove_unneeded_aliases = remove_unneeded_aliases
+        self._cross_validation_split_to_exclude = str(cross_validation_split_to_exclude)
+
+        self._sql_table_context = Text2SqlTableContext(schema_path)
+        self._world = Text2SqlWorld(self._sql_table_context)
+
+    @overrides
+    def _read(self, file_path: str):
+        """
+        This dataset reader consumes the data from
+        https://github.com/jkkummerfeld/text2sql-data/tree/master/data
+        formatted using ``scripts/reformat_text2sql_data.py``.
+
+        Parameters
+        ----------
+        file_path : ``str``, required.
+            For this dataset reader, file_path can either be a path to a file `or` a
+            path to a directory containing json files. The reason for this is because
+            some of the text2sql datasets require cross validation, which means they are split
+            up into many small files, for which you only want to exclude one.
+        """
+        files = [p for p in glob.glob(file_path)
+                 if self._cross_validation_split_to_exclude not in os.path.basename(p)]
+
+        for path in files:
+            with open(cached_path(path), "r") as data_file:
+                data = json.load(data_file)
+
+            for sql_data in text2sql_utils.process_sql_data(data,
+                                                            use_all_sql=self._use_all_sql,
+                                                            remove_unneeded_aliases=self._remove_unneeded_aliases,
+                                                            # TODO(Mark): Horrible hack, remove
+                                                            schema=self._sql_table_context.schema):
+                instance = self.text_to_instance(sql_data.text, sql_data)
+                if instance is not None:
+                    yield instance
+
+    @overrides
+    def text_to_instance(self,  # type: ignore
+                         query: List[str],
+                         sql: text2sql_utils.SqlData = None) -> Instance:
+        # pylint: disable=arguments-differ
+        fields: Dict[str, Field] = {}
+        tokens = TextField([Token(t) for t in query], self._token_indexers)
+        fields["tokens"] = tokens
+
+        if sql is not None:
+            try:
+                action_sequence, all_actions = self._world.get_action_sequence_and_all_actions(sql.sql)
+            except ParseError:
+                return None
+
+        index_fields: List[Field] = []
+        production_rule_fields: List[Field] = []
+
+        for production_rule in all_actions:
+            nonterminal, _ = production_rule.split(' ->')
+            # The whitespaces are not semantically meaningful, so we filter them out.
+            production_rule = ' '.join([token for token in production_rule.split(' ') if token != 'ws'])
```

Comment:
Do you need to worry about `wsp` here, too?

Reply:
Umm I think the production rule generation strips these out actually, so it never occurs. Let me double check that.